### PR TITLE
Cardinality detection and limiting implementation

### LINKED ISSUE #181: High cardinality detection to FFWD
High cardinality metrics are hard to manage and create unnecessary load on backend systems.
Earlier detection and following filtering of such metrics would help to automatically manage system without need for user intervention.

### PR DESCRIPTION
**Pull request by malish8632**
- Add cardinality limit based on [HyperLogLogPlus](https://github.com/addthis/stream-lib/blob/master/src/main/java/com/clearspring/analytics/stream/cardinality/HyperLogLogPlus.java)
- Add cardinality metric for observability
- For **batch** using "commonTags" and "commonResource" for hashing - feel this should be reviewed as we go to include all tags and values in the batch
- For single **metric** using same tags as original metric ("key", "riemannTags", "tags")
- Cardinality will be reset by swapping HyperLogLogPlus after configurable time in case it was tripped

Closes: #181

### CODE REVIEW

**Comment by lmuhlha** [comment]
seems reasonable to me 👍

File: core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java [hunk 1]
```diff
@@ -220,6 +292,22 @@ private boolean rateLimitAllowed(int permits, String key) {
...
     }
 
+    private boolean cardinalityLimitAllowed(long currentCardinality) {
+        return cardinalityLimit == null || cardinalityLimit >= currentCardinality;
+    }
+
+    /**
+    * To reset cardinality this will swap HLL++ if it was tripped after configured period of ms
+    */
+    private void swapHyperLogLogPlus() {
```
Comment (malish8632) [inline core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java:273]: @hexedpackets It looks ugly and I'm curious if there is other way to atomically swap the objects

File: core/src/main/java/com/spotify/ffwd/output/OutputManagerModule.java [hunk 2]
```diff
@@ -56,16 +56,22 @@
...
+    @Nullable private final Long hyperLogLogPlusSwapPeriod;
 
     @JsonCreator
     public OutputManagerModule(
         @JsonProperty("plugins") List<OutputPlugin> plugins,
         @JsonProperty("filter") Filter filter,
-        @JsonProperty("ratelimit") @Nullable Integer rateLimit
+        @JsonProperty("ratelimit") @Nullable Integer rateLimit,
+        @JsonProperty("cardinalitylimit") @Nullable Long cardinalityLimit,
+        @JsonProperty("hyperloglogswapperiod") @Nullable Long hyperLogLogPlusSwapPeriod
```
Comment (dmichel1) [inline core/src/main/java/com/spotify/ffwd/output/OutputManagerModule.java]: nit; include time unit in name.

File: core/pom.xml [hunk 3]
```diff
@@ -1,5 +1,17 @@
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
   <modelVersion>4.0.0</modelVersion>
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-compiler-plugin</artifactId>
```
Comment (dmichel1) [inline core/pom.xml]: did you mean to add this? why is this needed now. should be 11

File: core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java [hunk 4]
```diff
@@ -133,6 +161,22 @@ public final Long getRateLimit() {
         } else {
             rateLimiter = null;
         }
+
+        hyperLog = new HyperLogLogPlus(14, 25);
```
Comment (dmichel1) [inline core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java]: what are these magic numbers? would be good to define them as constants.

File: core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java [hunk 5]
```diff
@@ -176,10 +234,25 @@ public void sendBatch(Batch batch) {
...
         int batchSize = batch.getPoints().size();
 
-        if (batchSize > 0 && !rateLimitAllowed(batchSize, batch.getPoints().get(0).getKey())) {
-            log.debug("Dropping {} metrics due to rate limiting", batchSize);
-            statistics.reportMetricsDroppedByRateLimit(batchSize);
-            return;
+        hyperLog.offer(batch.hashCode());
+        statistics.reportMetricsCardinality(batch.hashCode());
+
+        if (!Arrays.asList(KEYS_NEVER_TO_DROP).contains(batch.getPoints().get(0).getKey())) {
```
Comment (dmichel1) [inline core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java]: this code for `batch` is nearly identical to the one for a single metric. can the code be deduplicated?

File: core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java [hunk 6]
```diff
@@ -220,6 +292,22 @@ private boolean rateLimitAllowed(int permits, String key) {
...
+        return cardinalityLimit == null || cardinalityLimit >= currentCardinality;
+    }
+
+    /**
+    * To reset cardinality this will swap HLL++ if it was tripped after configured period of ms
+    */
+    private void swapHyperLogLogPlus() {
+        if (System.currentTimeMillis() - hyperLogSwapTS.get() > hyperLogLogPlusSwapPeriod
+            && hyperLogSwapLock.compareAndExchange(false, true)) {
+          hyperLog = new HyperLogLogPlus(14, 25);
```
Comment (dmichel1) [inline core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java]: i might be wrong but I think you need a `synchronized` bit here so all threads will see the new reference. i need to come back to this part.

File: core/src/main/resources/log4j2.xml [hunk 7]
```diff
@@ -9,7 +9,7 @@
     <Logger name="kafka.consumer" level="INFO" additivity="false">
       <AppenderRef ref="primary" />
     </Logger>
-    <Logger name="com.spotify.ffwd" level="WARN" additivity="false">
+    <Logger name="com.spotify.ffwd" level="INFO" additivity="false">
```
Comment (dmichel1) [inline core/src/main/resources/log4j2.xml]: does this produce a lot of logs? iirc we changed this to reduce the logging in production. perhaps you can add a logger just for the class you are interested in instead?

File: core/pom.xml (hunk 3 above)
Comment (malish8632) [reply core/pom.xml]: nope, going to remove. Intellij gives me a headache with those modules and correct versions

File: core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java (hunk 4 above)
Comment (malish8632) [reply core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java]: done

File: core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java [hunk 8]
```diff
@@ -113,18 +118,41 @@ public long getTtl() {
...
             return null;
         }
         return rateLimiter.getCapacity();
     }
 
+    public final Long getCardinalityLimit() {
+        if (cardinalityLimit == null) {
+            return null;
+        }
+        return cardinalityLimit;
```
Comment (hexedpackets) [inline core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java]: This is redundant, you can just return `cardinalityLimit` without the null check.

File: core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java [hunk 9]
```diff
@@ -113,18 +118,41 @@ public long getTtl() {
...
+    }
+
+    public final Long getHLLPRefreshPeriodLimit() {
+        if (hyperLogLogPlusSwapPeriod == null) {
+            return null;
+        }
+        return hyperLogLogPlusSwapPeriod;
```
Comment (hexedpackets) [inline core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java]: Same here

File: core/src/main/java/com/spotify/ffwd/output/OutputManagerModule.java (hunk 2 above)
Comment (malish8632) [reply core/src/main/java/com/spotify/ffwd/output/OutputManagerModule.java]: done

File: core/src/main/resources/log4j2.xml (hunk 7 above)
Comment (malish8632) [reply core/src/main/resources/log4j2.xml]: reverting it back, just missed it

File: core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java (hunk 1 above)
Comment (hexedpackets) [reply core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java:273]: hmm, I can't think of a cleaner way to do it

Comment (malish8632) [reply core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java:273]: 👍

File: core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java (hunk 8 above)
Comment (malish8632) [reply core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java]: removed

File: core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java (hunk 5 above)
Comment (malish8632) [reply core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java]: i've consolidated it

File: core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java (hunk 6 above)
Comment (malish8632) [reply core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java]: I've updated this to use AtomicReference instead of an object itself.

File: agent/src/test/java/com/spotify/ffwd/FfwdConfigurationTest.java [hunk 10]
```diff
@@ -89,11 +89,15 @@ public void testConfMixedPluginsEnabled() {
     public void testConfigFromEnvVars() {
         environmentVariables.set("FFWD_TTL", "100");
         environmentVariables.set("FFWD_OUTPUT_RATELIMIT", "1000");
+        environmentVariables.set("FFWD_OUTPUT_CARDINALITYLIMIT", "10000");
+        environmentVariables.set("FFWD_OUTPUT_HYPERLOGLOGSWAPPERIODMS", "3000000");
```
Comment (dmichel1) [inline agent/src/test/java/com/spotify/ffwd/FfwdConfigurationTest.java]: i'm having a hard time reading `HYPERLOGLOGSWAPPERIODMS` can that be made more readable?

**Review by dmichel1 — APPROVED** [review]
nice - one more nit about naming then lgtm

File: agent/src/test/java/com/spotify/ffwd/FfwdConfigurationTest.java (hunk 10 above)
Comment (malish8632) [reply agent/src/test/java/com/spotify/ffwd/FfwdConfigurationTest.java]: done
